compiler/rustc_resolve/src/late.rs

```diff
@@ -2401,32 +2401,70 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
                 });
             }
 
-            ItemKind::Static(box ast::StaticItem { ref ty, ref expr, .. })
-            | ItemKind::Const(box ast::ConstItem { ref ty, ref expr, .. }) => {
+            ItemKind::Static(box ast::StaticItem { ref ty, ref expr, .. }) => {
                 self.with_static_rib(|this| {
                     this.with_lifetime_rib(LifetimeRibKind::Elided(LifetimeRes::Static), |this| {
                         this.visit_ty(ty);
                     });
                     this.with_lifetime_rib(LifetimeRibKind::Elided(LifetimeRes::Infer), |this| {
                         if let Some(expr) = expr {
-                            let constant_item_kind = match item.kind {
-                                ItemKind::Const(..) => ConstantItemKind::Const,
-                                ItemKind::Static(..) => ConstantItemKind::Static,
-                                _ => unreachable!(),
-                            };
                             // We already forbid generic params because of the above item rib,
                             // so it doesn't matter whether this is a trivial constant.
                             this.with_constant_rib(
                                 IsRepeatExpr::No,
                                 ConstantHasGenerics::Yes,
-                                Some((item.ident, constant_item_kind)),
+                                Some((item.ident, ConstantItemKind::Static)),
                                 |this| this.visit_expr(expr),
                             );
                         }
                     });
                 });
             }
 
+            ItemKind::Const(box ast::ConstItem { ref generics, ref ty, ref expr, .. }) => {
+                // FIXME(generic_consts): Is the ordering of ribs correct? Right now, it's
+                // STATIC <- LIFETIME_ANON_REPORT <- GENERICS <- LIFETIME_ELIDED_STATIC {,<- CONSTANT}
+                self.with_static_rib(|this| {
+                    this.with_lifetime_rib(LifetimeRibKind::AnonymousReportError, |this| {
```

Comment:
You should not need those two.

Reply:
Huh, that's curious! You seem to be right though, at least all tests in `tests/ui/` still pass after (locally) removing these ribs. As you can tell I'm not super familiar with ribs in rustc yet. I thought it was vital that `with_static_rib` sets up the `ValueNS` and `TypeNS` ribs.

Reply:
This is vital if nothing else does. Here, `with_generic_param_rib` sets up the three ribs `ValueNs`, `TypeNs` and lifetimes, so you don't need to add an extra layer.